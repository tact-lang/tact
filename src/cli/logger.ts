export const CliLogger = () => {
    let hadErrors = false;

    const log = (message: string) => {
        hadErrors = true;
        console.log(message);
    };

    return {
        log,
        hadErrors: () => hadErrors,
    };
};
