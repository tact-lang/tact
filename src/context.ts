export class CompilerContext {

    readonly shared: { [key: symbol]: object } = {};

    constructor(args: { shared: { [key: symbol]: object } } = { shared: {} }) {
        this.shared = args.shared;
        Object.freeze(this.shared);
        Object.freeze(this);
    }

    addShared = <T>(store: symbol, key: string | number, value: T) => {
        let sh: { [key: string]: T } = {};
        if (this.shared[store]) {
            sh = { ...this.shared[store] };
        }
        sh[key] = value;
        return new CompilerContext({ shared: { ...this.shared, [store]: sh } });
    }
}

export function createContextStore<T>() {
    const symbol = Symbol();
    return {
        get(ctx: CompilerContext, key: string | number) {
            if (!ctx.shared[symbol]) {
                return null;
            }
            const m = ctx.shared[symbol] as { [key: string | number]: T };
            if (m[key]) {
                return m[key];
            } else {
                return null;
            }
        },
        all(ctx: CompilerContext): { [key: string | number]: T } {
            if (!ctx.shared[symbol]) {
                return {} as { [key: string | number]: T };
            }
            const m = ctx.shared[symbol] as { [key: string | number]: T };
            return m;
        },
        set(ctx: CompilerContext, key: string | number, v: T) {
            return ctx.addShared(symbol, key, v);
        }
    }
}