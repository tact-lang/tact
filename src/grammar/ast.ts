import { Interval as RawInterval, Node as RawNode } from 'ohm-js';
import { TactSyntaxError } from '../errors';
import { TypeOrigin } from '../types/types';

export class ASTRef {

    static merge(...refs: ASTRef[]) {
        if (refs.length === 0) {
            throw Error('Cannot merge 0 refs');
        }
        let r = refs[0].#interval;
        const file = refs[0].#file;
        for (let i = 1; i < refs.length; i++) {
            r = r.coverageWith(r, refs[i].#interval);
        }
        return new ASTRef(r, file);
    }

    readonly #interval: RawInterval;
    readonly #file: string | null;

    constructor(interval: RawInterval, file: string | null) {
        this.#interval = interval;
        this.#file = file;
    }

    get file() {
        return this.#file;
    }

    get contents() {
        return this.#interval.contents;
    }

    get interval() {
        return this.#interval;
    }
}

export type ASTPrimitive = {
    kind: 'primitive',
    origin: TypeOrigin,
    id: number,
    name: string,
    ref: ASTRef
}

//
// Values
//

export type ASTNumber = {
    kind: 'number',
    id: number,
    value: bigint,
    ref: ASTRef
}

export type ASTID = {
    kind: 'id',
    id: number,
    value: string,
    ref: ASTRef
}

export type ASTBoolean = {
    kind: 'boolean',
    id: number,
    value: boolean,
    ref: ASTRef
}

export type ASTString = {
    kind: 'string',
    id: number,
    value: string,
    ref: ASTRef
}

export type ASTNull = {
    kind: 'null',
    id: number,
    ref: ASTRef
}

export type ASTLvalueRef = {
    kind: 'lvalue_ref',
    id: number,
    name: string,
    ref: ASTRef
}

//
// Types
//

export type ASTTypeRefSimple = {
    kind: 'type_ref_simple',
    id: number,
    name: string,
    optional: boolean,
    ref: ASTRef
}

export type ASTTypeRefMap = {
    kind: 'type_ref_map',
    id: number,
    key: string,
    keyAs: string | null,
    value: string,
    valueAs: string | null,
    ref: ASTRef
}

export type ASTTypeRefBounced = {
    kind: 'type_ref_bounced',
    id: number,
    name: string,
    ref: ASTRef
}

export type ASTTypeRef = ASTTypeRefSimple | ASTTypeRefMap | ASTTypeRefBounced;

//
// Expressions
//

export type ASTOpBinary = {
    kind: 'op_binary',
    id: number,
    op: '+' | '-' | '*' | '/' | '!=' | '>' | '<' | '>=' | '<=' | '==' | '&&' | '||' | '%' | '<<' | '>>' | '&' | '|',
    left: ASTExpression,
    right: ASTExpression,
    ref: ASTRef
}

export type ASTOpUnary = {
    kind: 'op_unary',
    id: number,
    op: '+' | '-' | '!' | '!!',
    right: ASTExpression,
    ref: ASTRef
}

export type ASTOpField = {
    kind: 'op_field'
    id: number,
    src: ASTExpression,
    name: string,
    ref: ASTRef
}

export type ASTOpCall = {
    kind: 'op_call'
    id: number,
    src: ASTExpression,
    name: string,
    args: ASTExpression[],
    ref: ASTRef
}

export type ASTOpCallStatic = {
    kind: 'op_static_call'
    id: number,
    name: string,
    args: ASTExpression[],
    ref: ASTRef
}

export type ASTOpNew = {
    kind: 'op_new'
    id: number,
    type: string,
    args: ASTNewParameter[],
    ref: ASTRef
}

export type ASTNewParameter = {
    kind: 'new_parameter'
    id: number,
    name: string,
    exp: ASTExpression,
    ref: ASTRef
}

export type ASTInitOf = {
    kind: 'init_of'
    id: number,
    name: string,
    args: ASTExpression[],
    ref: ASTRef
}

export type ASTConditional = {
    kind: 'conditional'
    id: number,
    condition: ASTExpression,
    thenBranch: ASTExpression,
    elseBranch: ASTExpression,
    ref: ASTRef
}

//
// Program
//

export type ASTProgram = {
    kind: 'program',
    id: number,
    entries: (ASTStruct | ASTContract | ASTPrimitive | ASTFunction | ASTNativeFunction | ASTTrait | ASTProgramImport | ASTConstant)[]
}

export type ASTProgramImport = {
    kind: 'program_import',
    id: number,
    path: ASTString,
    ref: ASTRef
}

export type ASTStruct = {
    kind: 'def_struct',
    origin: TypeOrigin,
    id: number,
    name: string,
    message: boolean,
    prefix: number | null,
    fields: ASTField[],
    ref: ASTRef
}

export type ASTTrait = {
    kind: 'def_trait',
    origin: TypeOrigin,
    id: number,
    name: string,
    traits: ASTString[],
    attributes: ASTContractAttribute[],
    declarations: (ASTField | ASTFunction | ASTReceive | ASTConstant)[],
    ref: ASTRef
}

export type ASTField = {
    kind: 'def_field',
    id: number,
    name: string,
    type: ASTTypeRef,
    init: ASTExpression | null,
    as: string | null,
    ref: ASTRef
}

export type ASTConstant = {
    kind: 'def_constant',
    id: number,
    name: string,
    type: ASTTypeRef,
    value: ASTExpression | null,
    attributes: ASTConstantAttribute[],
    ref: ASTRef
}

export type ASTConstantAttribute =
    | { type: 'virtual', ref: ASTRef }
    | { type: 'overrides', ref: ASTRef }
    | { type: 'abstract', ref: ASTRef };

export type ASTContractAttribute = { type: 'interface', name: ASTString, ref: ASTRef };

export type ASTContract = {
    kind: 'def_contract',
    origin: TypeOrigin,
    id: number,
    name: string,
    traits: ASTString[],
    attributes: ASTContractAttribute[],
    declarations: (ASTField | ASTFunction | ASTInitFunction | ASTReceive | ASTConstant)[],
    ref: ASTRef
}

export type ASTArgument = {
    kind: 'def_argument',
    id: number,
    name: string,
    type: ASTTypeRef,
    ref: ASTRef
}

export type ASTFunctionAttribute =
    | { type: 'get', ref: ASTRef }
    | { type: 'mutates', ref: ASTRef }
    | { type: 'extends', ref: ASTRef }
    | { type: 'virtual', ref: ASTRef }
    | { type: 'abstract', ref: ASTRef }
    | { type: 'overrides', ref: ASTRef }
    | { type: 'inline', ref: ASTRef };

export type ASTFunction = {
    kind: 'def_function',
    origin: TypeOrigin,
    id: number,
    attributes: ASTFunctionAttribute[],
    name: string,
    return: ASTTypeRef | null,
    args: ASTArgument[],
    statements: ASTStatement[] | null,
    ref: ASTRef
}

export type ASTReceive = {
    kind: 'def_receive',
    id: number,
    selector: {
        kind: 'internal-simple', arg: ASTArgument
    } | {
        kind: 'internal-fallback'
    } | {
        kind: 'internal-comment', comment: ASTString
    } | {
        kind: 'bounce', arg: ASTArgument
    } | {
        kind: 'external-simple', arg: ASTArgument
    } | {
        kind: 'external-fallback'
    } | {
        kind: 'external-comment', comment: ASTString
    },
    statements: ASTStatement[],
    ref: ASTRef
}

export type ASTNativeFunction = {
    kind: 'def_native_function',
    origin: TypeOrigin,
    id: number,
    attributes: ASTFunctionAttribute[],
    name: string,
    nativeName: string,
    return: ASTTypeRef | null,
    args: ASTArgument[],
    ref: ASTRef
}

export type ASTInitFunction = {
    kind: 'def_init_function',
    id: number,
    args: ASTArgument[],
    statements: ASTStatement[],
    ref: ASTRef
}

//
// Statements
//

export type ASTStatementLet = {
    kind: 'statement_let',
    id: number,
    name: string,
    type: ASTTypeRef,
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTStatementReturn = {
    kind: 'statement_return',
    id: number,
    expression: ASTExpression | null,
    ref: ASTRef
}

export type ASTStatementExpression = {
    kind: 'statement_expression',
    id: number,
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTSTatementAssign = {
    kind: 'statement_assign',
    id: number,
    path: ASTLvalueRef[],
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTSTatementAugmentedAssign = {
    kind: 'statement_augmentedassign',
    id: number,
    op: '+' | '-' | '*' | '/' | '%',
    path: ASTLvalueRef[],
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTCondition = {
    kind: 'statement_condition',
    id: number,
    expression: ASTExpression,
    trueStatements: ASTStatement[],
    falseStatements: ASTStatement[] | null;
    elseif: ASTCondition | null,
    ref: ASTRef,
}

export type ASTStatementWhile = {
    kind: 'statement_while'
    id: number,
    condition: ASTExpression,
    statements: ASTStatement[],
    ref: ASTRef
}

export type ASTStatementUntil = {
    kind: 'statement_until'
    id: number,
    condition: ASTExpression,
    statements: ASTStatement[],
    ref: ASTRef
}

export type ASTStatementRepeat = {
    kind: 'statement_repeat'
    id: number,
    condition: ASTExpression,
    statements: ASTStatement[],
    ref: ASTRef
}


//
// Unions
//

export type ASTStatement = ASTStatementLet | ASTStatementReturn | ASTStatementExpression | ASTSTatementAssign | ASTSTatementAugmentedAssign | ASTCondition | ASTStatementWhile | ASTStatementUntil | ASTStatementRepeat;
export type ASTNode = ASTExpression | ASTProgram | ASTStruct | ASTField | ASTContract | ASTArgument | ASTFunction | ASTOpCall | ASTStatementLet | ASTStatementReturn | ASTProgram | ASTPrimitive | ASTOpCallStatic | ASTStatementExpression | ASTNativeFunction | ASTSTatementAssign | ASTSTatementAugmentedAssign | ASTOpNew | ASTNewParameter | ASTTypeRef | ASTNull | ASTCondition | ASTInitFunction | ASTStatementWhile | ASTStatementUntil | ASTStatementRepeat | ASTReceive | ASTLvalueRef | ASTString | ASTTrait | ASTProgramImport | ASTFunction | ASTNativeFunction | ASTInitOf | ASTString | ASTConstant;
export type ASTExpression = ASTOpBinary | ASTOpUnary | ASTOpField | ASTNumber | ASTID | ASTBoolean | ASTOpCall | ASTOpCallStatic | ASTOpNew | ASTNull | ASTLvalueRef | ASTInitOf | ASTString | ASTConditional;
export type ASTType = ASTPrimitive | ASTStruct | ASTContract | ASTTrait;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type DistributiveOmit<T, K extends keyof any> = T extends any
    ? Omit<T, K>
    : never;
let nextId = 1;
export function createNode(src: DistributiveOmit<ASTNode, 'id'>): ASTNode {
    return Object.freeze(Object.assign({ id: nextId++ }, src));
}
export function cloneASTNode<T extends ASTNode>(src: T): T {
    return { ...src, id: nextId++ };
}

export function __DANGER_resetNodeId() {
    nextId = 1;
}

let currentFile: string | null = null;

export function inFile<T>(path: string, callback: () => T) {
    currentFile = path;
    const r = callback();
    currentFile = null;
    return r;
}

export function createRef(s: RawNode, ...extra: RawNode[]): ASTRef {
    let i = s.source;
    if (extra.length > 0) {
        i = i.coverageWith(...extra.map((e) => e.source));
    }
    return new ASTRef(i, currentFile);
}

export function throwError(message: string, ref: ASTRef): never {
    if (ref.file) {
        const lc = ref.interval.getLineAndColumn() as { lineNum: number, colNum: number };
        throw new TactSyntaxError(ref.file + ':' + lc.lineNum + ':' + lc.colNum + ': ' + message + '\n' + ref.interval.getLineAndColumnMessage(), ref);
    } else {
        throw new TactSyntaxError(message + ref.interval.getLineAndColumnMessage(), ref);
    }
}

export function traverse(node: ASTNode, callback: (node: ASTNode) => void) {
    callback(node);

    //
    // Program
    //

    if (node.kind === 'program') {
        for (const e of node.entries) {
            traverse(e, callback);
        }
    }
    if (node.kind === 'def_contract') {
        for (const e of node.declarations) {
            traverse(e, callback);
        }
    }
    if (node.kind === 'def_struct') {
        for (const e of node.fields) {
            traverse(e, callback);
        }
    }
    if (node.kind === 'def_trait') {
        for (const e of node.declarations) {
            traverse(e, callback);
        }
    }

    //
    // Functions
    //

    if (node.kind === 'def_function') {
        for (const e of node.args) {
            traverse(e, callback);
        }
        if (node.statements) {
            for (const e of node.statements) {
                traverse(e, callback);
            }
        }
    }
    if (node.kind === 'def_init_function') {
        for (const e of node.args) {
            traverse(e, callback);
        }
        for (const e of node.statements) {
            traverse(e, callback);
        }
    }
    if (node.kind === 'def_receive') {
        for (const e of node.statements) {
            traverse(e, callback);
        }
    }
    if (node.kind === 'def_native_function') {
        for (const e of node.args) {
            traverse(e, callback);
        }
    }
    if (node.kind === 'def_field') {
        if (node.init) {
            traverse(node.init, callback);
        }
    }
    if (node.kind === 'def_constant') {
        if (node.value) {
            traverse(node.value, callback);
        }
    }

    //
    // Statements
    //

    if (node.kind === 'statement_let') {
        traverse(node.type, callback);
        traverse(node.expression, callback);
    }
    if (node.kind === 'statement_return') {
        if (node.expression) {
            traverse(node.expression, callback);
        }
    }
    if (node.kind === 'statement_expression') {
        traverse(node.expression, callback);
    }
    if (node.kind === 'statement_assign') {
        for (const e of node.path) {
            traverse(e, callback);
        }
        traverse(node.expression, callback);
    }
    if (node.kind === 'statement_augmentedassign') {
        for (const e of node.path) {
            traverse(e, callback);
        }
        traverse(node.expression, callback);
    }
    if (node.kind === 'statement_condition') {
        traverse(node.expression, callback);
        for (const e of node.trueStatements) {
            traverse(e, callback);
        }
        if (node.falseStatements) {
            for (const e of node.falseStatements) {
                traverse(e, callback);
            }
        }
        if (node.elseif) {
            traverse(node.elseif, callback);
        }
    }
    if (node.kind === 'statement_while') {
        traverse(node.condition, callback);
        for (const e of node.statements) {
            traverse(e, callback);
        }
    }
    if (node.kind === 'statement_until') {
        traverse(node.condition, callback);
        for (const e of node.statements) {
            traverse(e, callback);
        }
    }
    if (node.kind === 'statement_repeat') {
        traverse(node.condition, callback);
        for (const e of node.statements) {
            traverse(e, callback);
        }
    }
    if (node.kind === 'op_binary') {
        traverse(node.left, callback);
        traverse(node.right, callback);
    }
    if (node.kind === 'op_unary') {
        traverse(node.right, callback);
    }
    if (node.kind === 'op_field') {
        traverse(node.src, callback);
    }
    if (node.kind === 'op_call') {
        traverse(node.src, callback);
        for (const e of node.args) {
            traverse(e, callback);
        }
    }
    if (node.kind === 'op_static_call') {
        for (const e of node.args) {
            traverse(e, callback);
        }
    }
    if (node.kind === 'op_new') {
        for (const e of node.args) {
            traverse(e, callback);
        }
    }
    if (node.kind === 'new_parameter') {
        traverse(node.exp, callback);
    }
    if (node.kind === 'conditional') {
        traverse(node.condition, callback);
        traverse(node.thenBranch, callback);
        traverse(node.elseBranch, callback);
    }
}