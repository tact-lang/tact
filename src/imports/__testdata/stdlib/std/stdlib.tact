import "./stdlib2.fc";