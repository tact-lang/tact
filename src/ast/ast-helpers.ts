import * as A from "./ast";
import { AstId } from "./ast";
import { throwInternalCompilerError } from "../error/errors";
import { dummySrcInfo } from "../grammar";

/**
 * Check if input expression is a 'path expression',
 * i.e. an identifier or a sequence of field accesses starting from an identifier.
 * @param path A path expression to check.
 * @returns An array of identifiers or null if the input expression is not a path expression.
 */
export function tryExtractPath(path: A.AstExpression): A.AstId[] | null {
    switch (path.kind) {
        case "id":
            return [path];
        case "field_access": {
            const p = tryExtractPath(path.aggregate);
            return p ? [...p, path.field] : null;
        }
        default:
            return null;
    }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type DistributiveOmit<T, K extends keyof any> = T extends any
    ? Omit<T, K>
    : never;

export const getAstFactory = () => {
    let nextId = 1;
    function createNode(src: DistributiveOmit<A.AstNode, "id">): A.AstNode {
        return Object.freeze(Object.assign({ id: nextId++ }, src));
    }
    function cloneNode<T extends A.AstNode>(src: T): T {
        const newNode: T = { ...src, id: nextId++ };
        return Object.freeze(newNode);
    }
    return {
        createNode,
        cloneNode,
    };
};

export type FactoryAst = ReturnType<typeof getAstFactory>;

export function idText(ident: A.AstId | A.AstFuncId | A.AstTypeId): string {
    return ident.text;
}

export function isInt(ident: A.AstTypeId): boolean {
    return ident.text === "Int";
}

export function isBool(ident: A.AstTypeId): boolean {
    return ident.text === "Bool";
}

export function isCell(ident: A.AstTypeId): boolean {
    return ident.text === "Cell";
}

export function isSlice(ident: A.AstTypeId): boolean {
    return ident.text === "Slice";
}

export function isBuilder(ident: A.AstTypeId): boolean {
    return ident.text === "Builder";
}

export function isAddress(ident: A.AstTypeId): boolean {
    return ident.text === "Address";
}

export function isString(ident: A.AstTypeId): boolean {
    return ident.text === "String";
}

export function isStringBuilder(ident: A.AstTypeId): boolean {
    return ident.text === "StringBuilder";
}

export function isSelfId(ident: A.AstId): boolean {
    return ident.text === "self";
}

export function isWildcard(ident: A.AstId): boolean {
    return ident.text === "_";
}

export function isRequire(ident: A.AstId): boolean {
    return ident.text === "require";
}

export function eqNames(
    left: A.AstId | A.AstTypeId | string,
    right: A.AstId | A.AstTypeId | string,
): boolean {
    if (typeof left === "string") {
        if (typeof right === "string") {
            return left === right;
        }
        return left === right.text;
    } else {
        if (typeof right === "string") {
            return left.text === right;
        }
        return left.text === right.text;
    }
}

export function idOfText(text: string): A.AstId {
    return {
        kind: "id",
        text,
        id: 0,
        loc: dummySrcInfo,
    };
}

export function astNumToString(n: A.AstNumber): string {
    switch (n.base) {
        case 2:
            return `0b${n.value.toString(n.base)}`;
        case 8:
            return `0o${n.value.toString(n.base)}`;
        case 10:
            return n.value.toString(n.base);
        case 16:
            return `0x${n.value.toString(n.base)}`;
    }
}

// Test equality of AstExpressions.
// Note this is syntactical equality of expressions.
// For example, two struct instances are equal if they have the same
// type and same fields in the same order.
export function eqExpressions(
    ast1: A.AstExpression,
    ast2: A.AstExpression,
): boolean {
    if (ast1.kind !== ast2.kind) {
        return false;
    }

    switch (ast1.kind) {
        case "null":
            return true;
        case "boolean":
            return ast1.value === (ast2 as A.AstBoolean).value;
        case "number":
            return ast1.value === (ast2 as A.AstNumber).value;
        case "string":
            return ast1.value === (ast2 as A.AstString).value;
        case "id":
            return eqNames(ast1, ast2 as A.AstId);
        case "address":
            return ast1.value.equals((ast2 as A.AstAddress).value);
        case "cell":
            return ast1.value.equals((ast2 as A.AstCell).value);
        case "slice":
            return ast1.value
                .asCell()
                .equals((ast2 as A.AstSlice).value.asCell());
        case "simplified_string":
            return ast1.value === (ast2 as A.AstSimplifiedString).value;
        case "struct_value":
            return (
                eqNames(ast1.type, (ast2 as A.AstStructValue).type) &&
                eqArrays(
                    ast1.args,
                    (ast2 as A.AstStructValue).args,
                    eqFieldValues,
                )
            );
        case "method_call":
            return (
                eqNames(ast1.method, (ast2 as A.AstMethodCall).method) &&
                eqExpressions(ast1.self, (ast2 as A.AstMethodCall).self) &&
                eqArrays(
                    ast1.args,
                    (ast2 as A.AstMethodCall).args,
                    eqExpressions,
                )
            );
        case "init_of":
            return (
                eqNames(ast1.contract, (ast2 as A.AstInitOf).contract) &&
                eqArrays(ast1.args, (ast2 as A.AstInitOf).args, eqExpressions)
            );
        case "op_unary":
            return (
                ast1.op === (ast2 as A.AstOpUnary).op &&
                eqExpressions(ast1.operand, (ast2 as A.AstOpUnary).operand)
            );
        case "op_binary":
            return (
                ast1.op === (ast2 as A.AstOpBinary).op &&
                eqExpressions(ast1.left, (ast2 as A.AstOpBinary).left) &&
                eqExpressions(ast1.right, (ast2 as A.AstOpBinary).right)
            );
        case "conditional":
            return (
                eqExpressions(
                    ast1.condition,
                    (ast2 as A.AstConditional).condition,
                ) &&
                eqExpressions(
                    ast1.thenBranch,
                    (ast2 as A.AstConditional).thenBranch,
                ) &&
                eqExpressions(
                    ast1.elseBranch,
                    (ast2 as A.AstConditional).elseBranch,
                )
            );
        case "struct_instance":
            return (
                eqNames(ast1.type, (ast2 as A.AstStructInstance).type) &&
                eqArrays(
                    ast1.args,
                    (ast2 as A.AstStructInstance).args,
                    eqFieldInitializers,
                )
            );
        case "field_access":
            return (
                eqNames(ast1.field, (ast2 as A.AstFieldAccess).field) &&
                eqExpressions(
                    ast1.aggregate,
                    (ast2 as A.AstFieldAccess).aggregate,
                )
            );
        case "static_call":
            return (
                eqNames(ast1.function, (ast2 as A.AstStaticCall).function) &&
                eqArrays(
                    ast1.args,
                    (ast2 as A.AstStaticCall).args,
                    eqExpressions,
                )
            );
        default:
            throwInternalCompilerError("Unrecognized expression kind");
    }
}

function eqFieldInitializers(
    arg1: A.AstStructFieldInitializer,
    arg2: A.AstStructFieldInitializer,
): boolean {
    return (
        eqNames(arg1.field, arg2.field) &&
        eqExpressions(arg1.initializer, arg2.initializer)
    );
}

function eqFieldValues(
    arg1: A.AstStructFieldValue,
    arg2: A.AstStructFieldValue,
): boolean {
    return (
        eqNames(arg1.field, arg2.field) &&
        eqExpressions(arg1.initializer, arg2.initializer)
    );
}

function eqArrays<T>(
    arr1: readonly T[],
    arr2: readonly T[],
    eqElements: (elem1: T, elem2: T) => boolean,
): boolean {
    if (arr1.length !== arr2.length) {
        return false;
    }

    for (let i = 0; i < arr1.length; i++) {
        if (!eqElements(arr1[i]!, arr2[i]!)) {
            return false;
        }
    }

    return true;
}

/*
Functions that return guard types like "ast is AstLiteral" are unsafe to use in production code.
But there is a way to make them safe by introducing an intermediate function, like
the "checkLiteral" function defined below after "isLiteral". In principle, it is possible to use "checkLiteral"
directly in the code (which avoids the guard type altogether), but it produces code that reduces readability significantly.

The pattern shown with "isLiteral" and "checkLiteral" can be generalized to other functions that produce a guard type
based on a decision of several cases.
For example, if we have the following function, where we assume that B is a subtype of A:

function isB(d: A): d is B {
  if (cond1(d)) {        // It is assumed that cond1(d) determines d to be of type B inside the if
     return true;
  } else if (cond2(d)) { // It is assumed that cond2(d) determines d to be of type A but not of type B inside the if
     return false;
  } else if (cond3(d)) { // It is assumed that cond3(d) determines d to be of type B inside the if
     return true;
  } else {               // It is assumed that d is of type A but not of type B inside the else
     return false;
  }
}

We can introduce a "checkB" function as follows:

function checkB<T>(d: A, t: (arg: B) => T, f: (arg: Exclude<A,B>) => T): T {
  if (cond1(d)) {
     return t(d);
  } else if (cond2(d)) {
     return f(d);
  } else if (cond3(d)) {
     return t(d);
  } else {
     return f(d);
  }
}

Here, all the "true" cases return t(d) and all the "false" cases return f(d). The names of the functions t and f help remember
that they correspond to the true and false cases, respectively. Observe that cond1(d) and cond3(d) determine the type of
d to be B, which means we can pass d to the t function. For the false cases, the type of d is determined to be
A but not B, which means we can pass d to function f, because f's argument type Exclude<A,B> states
that the argument must be of type A but not of type B, i.e., of type "A - B" if we see the types as sets.

checkB is safe because the compiler will complain if, for example, we use t(d) in the else case:

function checkB<T>(d: A, t: (arg: B) => T, f: (arg: Exclude<A,B>) => T): T {
  if (cond1(d)) {
     return t(d);
  } else if (cond2(d)) {
     return f(d);
  } else if (cond3(d)) {
     return t(d);
  } else {
     return t(d);   // Compiler will signal an error that d is not assignable to type B
  }
}

Contrary to the original function, where the compiler remains silent if we incorrectly return true in the else:

function isB(d: A): d is B {
  if (cond1(d)) {
     return true;
  } else if (cond2(d)) {
     return false;
  } else if (cond3(d)) {
     return true;
  } else {
     return true;   // Wrong, but compiler remains silent
  }
}

After we have our "checkB" function, we can define the "isB" function simply as:

function isB(d: A): d is B {
  return checkB(d, () => true, () => false);
}
*/

export function isLiteral(ast: A.AstExpression): ast is A.AstLiteral {
    return checkLiteral(
        ast,
        () => true,
        () => false,
    );
}

function checkLiteral<T>(
    ast: A.AstExpression,
    t: (node: A.AstLiteral) => T,
    f: (node: Exclude<A.AstExpression, A.AstLiteral>) => T,
): T {
    switch (ast.kind) {
        case "null":
        case "boolean":
        case "number":
        case "address":
        case "cell":
        case "slice":
        case "simplified_string":
        case "struct_value":
            return t(ast);

        case "struct_instance":
        case "string":
        case "id":
        case "method_call":
        case "init_of":
        case "op_unary":
        case "op_binary":
        case "conditional":
        case "field_access":
        case "static_call":
            return f(ast);

        default:
            throwInternalCompilerError("Unrecognized expression kind");
    }
}

export const selfId: AstId = {
    kind: "id",
    text: "self",
    id: 0,
    loc: dummySrcInfo,
};
