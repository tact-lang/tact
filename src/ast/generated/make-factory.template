// THIS IS AN AUTOGENERATED FILE. TO GENERATE IT AGAIN, EXECUTE yarn gen:make-funs

import type { FactoryAst } from "@/ast/ast-helpers";
import type * as Ast from "@/ast/ast";
import type { SrcInfo } from "@/grammar";
import type * as TonCore from "@ton/core";
import { getSrcInfo } from "@/grammar/src-info";

export const getMakeAst = (astF: FactoryAst) => {
    
    const emptySrcInfo = getSrcInfo(" ", 0, 0, null, "user");
    
    <FUNCTIONS>

    return {
        <FUNCTION_NAMES>
    };
};

export type MakeAstFactory = ReturnType<typeof getMakeAst>;