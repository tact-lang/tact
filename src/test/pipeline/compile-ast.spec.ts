import type * as Ast from "@/ast/ast";
import { getAstFactory } from "@/ast/ast-helpers";
import { CompilerContext } from "@/context/context";
import { precompile } from "@/pipeline/precompile";
import * as Stdlib from "@/stdlib/stdlib";
import { createVirtualFileSystem } from "@/vfs/createVirtualFileSystem";
import type { MakeAstFactory } from "@/ast/generated/make-factory";
import { getMakeAst } from "@/ast/generated/make-factory";
import type { BuildContext } from "@/pipeline/build";
import { Logger, LogLevel } from "@/context/logger";
import { compileFunc, compileTact } from "@/pipeline/compile";
import { AssemblyWriter, disassembleRoot } from "@tact-lang/opcode";
import { Cell } from "@ton/core";
import { getType } from "@/types/resolveDescriptors";

// This function manually creates a module containing a contract with a reference to StateInit in its single field.
// StateInit is declared in stdlib.
//
// contract Test {
//    f: StateInit;
//
//    init() {
//        self.f = initOf Test();
//    }
// }

function makeModule(mF: MakeAstFactory): Ast.Module {
    // The contract field
    const field = mF.makeDummyFieldDecl(
        mF.makeDummyId("f"),
        mF.makeDummyTypeId("StateInit"),
        undefined,
        undefined,
    );
    // The contract init function
    const initOfExpr = mF.makeDummyInitOf(mF.makeDummyId("Test"), []);
    const path = mF.makeDummyFieldAccess(
        mF.makeDummyId("self"),
        mF.makeDummyId("f"),
    );
    const assignStmt = mF.makeDummyStatementAssign(path, initOfExpr);
    const init = mF.makeDummyContractInit([], [assignStmt]);

    // The contract
    const contract = mF.makeDummyContract(
        mF.makeDummyId("Test"),
        [],
        [],
        undefined,
        [field, init],
    );

    return mF.makeModule([], [contract]);
}

describe("compilation of ASTs", () => {
    const astF = getAstFactory();
    const mF = getMakeAst(astF);

    it("should compile AST to BoC", async () => {
        const moduleAst = makeModule(mF);
        const codeBoc = await compileModule(moduleAst, "Test");

        if (!codeBoc) {
            fail("Cannot compile FunC");
        }

        const codeFiftDecompiled = decompileCell(codeBoc);

        expect(codeBoc.toString("hex")).toMatchSnapshot();
        expect(codeFiftDecompiled).toMatchSnapshot();
    });

    async function compileModule(moduleAst: Ast.Module, contractName: string) {
        const fileSystem = { ["dummy.tact"]: "" };

        const project = createVirtualFileSystem("/", fileSystem, false);
        const stdlib = createVirtualFileSystem("@stdlib", Stdlib.files);

        let ctx = new CompilerContext();
        ctx = precompile(ctx, project, stdlib, "dummy.tact", [moduleAst]);

        const bCtx: BuildContext = {
            config: {
                name: "test",
                path: "",
                mode: "full",
                output: "./out",
            },
            logger: new Logger(LogLevel.NONE),
            project,
            stdlib,
            ctx,
            compilerInfo: "",
            built: {},
            errorMessages: [],
        };

        const contract = getType(ctx, contractName);
        const compileRes = await compileTact(bCtx, contract);
        if (!compileRes) {
            fail(`Tact compiler failed`);
        }

        return await compileFunc(
            bCtx,
            contractName,
            compileRes.entrypointPath,
            compileRes.funcSource,
        );
    }

    function decompileCell(codeBoc: Buffer) {
        const cell = Cell.fromBoc(codeBoc).at(0);
        if (typeof cell === "undefined") {
            fail("Cannot create Cell from BoC file");
        }

        const program = disassembleRoot(cell, { computeRefs: true });
        return AssemblyWriter.write(program, {
            useAliases: true,
        });
    }
});
