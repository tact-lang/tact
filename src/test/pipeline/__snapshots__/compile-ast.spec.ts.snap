// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compilation of ASTs should compile AST to BoC 1`] = `"b5ee9c72010102010045000114ff00f4a413f4bcf2c80b01006cd330d072d721d200d200fa4021103450556f04f86101f862ed44d0d2000195d4d4596c129b30f82ac87001ca006d30c9e25bdcf2c082"`;

exports[`compilation of ASTs should compile AST to BoC 2`] = `
""Asm.fif" include
PROGRAM{
  DECLPROC recv_internal
  recv_internal PROC:<{
    DROP
    CTOS
    TWO
    SDSKIPFIRST
    1 LDI
    1 LDI
    LDMSGADDR
    OVER
    s3 s4 XCHG
    s5 s5 XCHG2
    4 TUPLE
    1 SETGLOB
    SWAP
    2 SETGLOB
    PUSHROOT
    CTOS
    1 LDI
    SWAP
    <{
      LDREF
      LDREF
      ROTREV
      1 2 BLKDROP2
    }> PUSHCONT
    <{
      DROP
      MYCODE
      NEWC
      ZERO
      SWAP
      1 STI
      NULL
      DROP
      ENDC
    }> PUSHCONT
    IFELSE
    DROP2
    IFRET
    130 THROW
  }>
}END>c"
`;
